File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/CustomItemStackHandlerCodec.java
package com.gregtechceu.gtceu.api.sync_system.codecs;

import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec;
import com.gregtechceu.gtceu.api.transfer.item.CustomItemStackHandler;
import com.gregtechceu.gtceu.common.data.GTDataComponents;
import com.gregtechceu.gtceu.common.data.datacomponents.TransferData;

import net.minecraft.core.component.DataComponentMap;

import com.google.gson.JsonElement;
import com.mojang.serialization.JsonOps;

public final class CustomItemStackHandlerCodec implements ContextualFieldCodec<CustomItemStackHandler> {

    public static final Class<CustomItemStackHandler> TYPE = CustomItemStackHandler.class;
    public static final CustomItemStackHandlerCodec INSTANCE = new CustomItemStackHandlerCodec();

    private CustomItemStackHandlerCodec() {}

    @Override
    public JsonElement serializeField(CustomItemStackHandler value, Context<CustomItemStackHandler> context) {
        return DataComponentMap.CODEC
                .encodeStart(context.lookup().createSerializationContext(JsonOps.INSTANCE), value.exportComponents())
                .getOrThrow();
    }

    @Override
    public CustomItemStackHandler deserializeField(JsonElement value, Context<CustomItemStackHandler> context) {
        DataComponentMap components = DataComponentMap.CODEC
                .parse(context.lookup().createSerializationContext(JsonOps.INSTANCE), value)
                .getOrThrow();
        TransferData.ItemHandler data = components.get(GTDataComponents.TRANSFER_ITEM_HANDLER.get());
        if (data == null) {
            throw new IllegalArgumentException("Sync: item handler field " + context.fieldName() +
                    " is missing transfer_item_handler");
        }
        CustomItemStackHandler handler = context.currentValue();
        if (handler == null) {
            handler = new CustomItemStackHandler(data.slots());
        }

        handler.importComponents(components);
        return handler;
    }
}

File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/CustomItemStackHandlerCodec.kt
package com.gregtechceu.gtceu.api.sync_system.codecs

import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec
import com.gregtechceu.gtceu.api.transfer.item.CustomItemStackHandler
import com.gregtechceu.gtceu.common.data.GTDataComponents

import net.minecraft.core.component.DataComponentMap

import com.google.gson.JsonElement
import com.mojang.serialization.JsonOps

class CustomItemStackHandlerCodec private constructor() : ContextualFieldCodec<CustomItemStackHandler> {

	override fun serializeField(value: CustomItemStackHandler, context: ContextualFieldCodec.Context<CustomItemStackHandler>): JsonElement = DataComponentMap.CODEC
		.encodeStart(context.lookup.createSerializationContext(JsonOps.INSTANCE), value.exportComponents())
		.getOrThrow()

	override fun deserializeField(value: JsonElement, context: ContextualFieldCodec.Context<CustomItemStackHandler>): CustomItemStackHandler {
		val components = DataComponentMap.CODEC
			.parse(context.lookup.createSerializationContext(JsonOps.INSTANCE), value)
			.getOrThrow()
		val data = components.get(GTDataComponents.TRANSFER_ITEM_HANDLER.get())
			?: throw IllegalArgumentException("Sync: item handler field ${context.fieldName} is missing transfer_item_handler")
		val handler = context.currentValue ?: CustomItemStackHandler(data.slots)
		handler.importComponents(components)
		return handler
	}

	companion object {
		@JvmField
		val TYPE: Class<CustomItemStackHandler> = CustomItemStackHandler::class.java

		@JvmField
		val INSTANCE = CustomItemStackHandlerCodec()
	}
}
